src/improved_gmres.jl

```diff
@@ -0,0 +1,211 @@
+        end
+
+        # We already have the initial residual
+        if restarts > 1
+
+            # Set the first basis vector
+            β = init!(arnoldi, x, b, Pl, reserved_vec)
+
+            # And initialize the residual
+            init_residual!(residual, β)
+            
+            if log
+                history.mvps += 1
+            end
+        end
+
+        # Inner iterations k = 1, ..., restart
+        k = 1
+
+        while k ≤ restart && residual.current > reltol
+
+            # Arnoldi step: expand and orthogonalize
+            expand!(arnoldi, Pl, Pr, k)
+            orthogonalize!(arnoldi, k)
+            update_residual!(residual, arnoldi, k)
+            
+            if log
+                nextiter!(history, mvps = 1)
+                push!(history, :resnorm, residual.current)
+            end
+            
+            verbose && @printf("%3d\t%3d\t%1.2e\n", restarts, k, residual.current)
+
+            k += 1
+        end
+
+        # Solve the projected problem Hy = β * e1 in the least-squares sense
+        y = solve_least_squares!(arnoldi, β, k)
+
+        # And improve the solution x ← x + Pr \ (V * y)
+        update_solution!(x, y, arnoldi, Pr, k)
+    end
+
+    verbose && @printf("\n")
+    x
+end
+
+mutable struct ArnoldiDecomp{T}
+    A
+    V::Vector{Vector{T}} # Orthonormal basis vectors
+    H::Matrix{T}         # Hessenberg matrix
+end
+
+ArnoldiDecomp(A, order::Int, T::Type) = ArnoldiDecomp{T}(
+    A,
+    [zeros(T, size(A, 1)) for i = 1 : order + 1],
+    zeros(T, order + 1, order)
+)
+
+mutable struct Residual{numT, resT}
+    current::resT # Current relative residual
+    accumulator::resT # Used to compute the residual on the go
+    nullvec::Vector{numT} # Vector in the null space to compute residuals
+    β::resT # the initial residual
+end
+
+Residual(order, T::Type) = Residual{T, real(T)}(
+    one(real(T)),
+    one(real(T)),
+    ones(T, order + 1),
+    one(real(T))
+)
+
+function update_residual!(r::Residual, arnoldi::ArnoldiDecomp, k::Int)
+    # Cheaply computes the current residual
+    r.nullvec[k + 1] = -conj(dot(UnsafeVectorView(r.nullvec, 1 : k), UnsafeVectorView(arnoldi.H, 1 : k, k)) / arnoldi.H[k + 1, k])
+    r.accumulator += abs2(r.nullvec[k + 1])
+    r.current = r.β / √r.accumulator
+end
+
+function init!(arnoldi::ArnoldiDecomp{T}, x, b, Pl, reserved_vec) where {T}
+    # Initialize the Krylov subspace with the initial residual vector
+    # This basically does V[1] = Pl \ (b - A * x) and then normalize
+    copy!(arnoldi.V[1], b)
+    A_mul_B!(reserved_vec, arnoldi.A, x)
+    @blas! arnoldi.V[1] -= one(T) * reserved_vec
+    A_ldiv_B!(Pl, arnoldi.V[1])
+
+    # Normalize
+    β = norm(arnoldi.V[1])
+    @blas! arnoldi.V[1] *= one(T) / β
+    β
+end
+
+@inline function init_residual!(r::Residual{numT, resT}, β) where {numT,resT}
+    r.accumulator = one(resT)
+    r.β = β
+end
+
+function solve_least_squares!(arnoldi::ArnoldiDecomp{T}, β, k::Int) where {T}
```

Comment:
Maybe a more precise name here since it is not solving arbitrary least squares problems